lib/AST/LayoutConstraint.cpp

```diff
@@ -114,4 +143,74 @@ void *LayoutConstraintInfo::operator new(size_t bytes, const ASTContext &ctx,
 
 SourceRange LayoutConstraintLoc::getSourceRange() const { return getLoc(); }
 
+#define MERGE_TO(lhs, rhs, result)                                             \
+  mergeTable[int(LayoutConstraintKind::lhs)][int(LayoutConstraintKind::rhs)] = \
+      LayoutConstraintKind::result
+
+#define MERGE_LOOKUP(lhs, rhs)                                                 \
+  mergeTable[int(lhs)][int(rhs)]
+
+LayoutConstraint LayoutConstraint::merge(LayoutConstraint Other) {
+  /// Define a mapping
+  // (LayoutConstraintKind, LayoutConstraintKind) -> LayoutConstraintKind
+  // It shows what kind of layout constraint would be produced by merging of two
+  // layout constraints with given kinds.
+  static LayoutConstraintKind
+      mergeTable[unsigned(LayoutConstraintKind::LastLayout) + 1]
+                [unsigned(LayoutConstraintKind::LastLayout) + 1];
+  static int initializedMergeTable;
+
+  if (!initializedMergeTable) {
+    // initialize the merge table once.
+    initializedMergeTable = true;
+    MERGE_TO(TrivialOfAtMostSize, Trivial, TrivialOfAtMostSize);
+    MERGE_TO(TrivialOfExactSize, Trivial, TrivialOfExactSize);
+    MERGE_TO(Trivial, TrivialOfAtMostSize, TrivialOfAtMostSize);
+    MERGE_TO(Trivial, TrivialOfExactSize, TrivialOfExactSize);
+
+    MERGE_TO(NativeRefCountedObject, RefCountedObject, NativeRefCountedObject);
+    MERGE_TO(NativeRefCountedObject, Class, NativeRefCountedObject);
+    MERGE_TO(NativeClass, RefCountedObject, NativeClass);
+    MERGE_TO(NativeClass, Class, NativeClass);
+
+    MERGE_TO(RefCountedObject, NativeRefCountedObject, NativeRefCountedObject);
+    MERGE_TO(Class, NativeRefCountedObject, NativeRefCountedObject);
+    MERGE_TO(RefCountedObject, NativeClass, NativeClass);
+    MERGE_TO(Class, NativeClass, NativeClass);
+  }
+
+  // Lookup in the mergeTable if this combination of layouts can be merged.
+  auto mergeKind = MERGE_LOOKUP((*this)->getKind(), Other->getKind());
+
+  if (mergeKind == LayoutConstraintKind::UnknownLayout)
+    return LayoutConstraint::getUnknownLayout();
+
+  if ((*this)->getKind() == mergeKind)
+    return *this;
```

Comment:
When the result is one of the sized cases (`TrivialOfExactSize` or `TrivialOfAtMostSize`), you'll need to do some checking of the size to determine the proper result. The lookup table doesn't always suffice in these cases... for example, the `merge(TrivialOfExactSize, TrivialOfExactSize)` case might produce `TrivialOfExactSize` (if the sizes are the same) and might produce `TrivialOfAtMostSize` or `UnknownLayout` (if the sizes are different; we can choose which semantics we prefer).